test/downloadutils.mjs

```diff
@@ -25,89 +24,74 @@ function rewriteWebArchiveUrl(url) {
   // Web Archive URLs need to be transformed to add `if_` after the ID.
   // Without this, an HTML page containing an iframe with the PDF file
   // will be served instead (issue 8920).
-  var webArchiveRegex =
+  const webArchiveRegex =
     /(^https?:\/\/web\.archive\.org\/web\/)(\d+)(\/https?:\/\/.+)/g;
-  var urlParts = webArchiveRegex.exec(url);
+  const urlParts = webArchiveRegex.exec(url);
   if (urlParts) {
-    return urlParts[1] + (urlParts[2] + "if_") + urlParts[3];
+    return `${urlParts[1]}${urlParts[2]}if_${urlParts[3]}`;
   }
   return url;
 }
 
-function downloadFile(file, url, callback, redirects) {
+function downloadFile(file, url, redirects) {
   url = rewriteWebArchiveUrl(url);
+  const protocol = /^https:\/\//.test(url) ? https : http;
 
-  var protocol = /^https:\/\//.test(url) ? https : http;
-  protocol
-    .get(url, function (response) {
-      if (
-        response.statusCode === 301 ||
-        response.statusCode === 302 ||
-        response.statusCode === 307 ||
-        response.statusCode === 308
-      ) {
-        if (redirects > 10) {
-          callback("Too many redirects");
+  return new Promise((resolve, reject) => {
+    protocol
+      .get(url, async function (response) {
+        if ([301, 302, 307, 308].includes(response.statusCode)) {
+          if (redirects > 10) {
+            reject(new Error("Too many redirects"));
+            return;
+          }
+          const redirectTo = urlResolve(url, response.headers.location);
+          try {
+            await downloadFile(file, redirectTo, (redirects || 0) + 1);
```

Comment:
Assuming the comment above is implemented, this could be shortened a little bit.
```suggestion
            await downloadFile(file, redirectTo, ++redirects);
```